File: apps/ios/SocialApp/VoiceApi.swift
// Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
import Foundation
import SwiftUI

struct CallRecordItem: Codable, Identifiable {
    let id: Int64
    let callerId: Int64
    let calleeId: Int64
    let status: Int
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id
        case callerId = "caller_id"
        case calleeId = "callee_id"
        case status
        case createdAt = "created_at"
    }
}

struct VoiceRoomItem: Codable, Identifiable {
    let id: Int64
    let ownerId: Int64
    let name: String
    let onlineCount: Int
    let micCount: Int

    enum CodingKeys: String, CodingKey {
        case id
        case ownerId = "owner_id"
        case name
        case onlineCount = "online_count"
        case micCount = "mic_count"
    }
}

/// 语音 API。所有请求带 Authorization + X-Api-Version 头；
/// 语音文件 GET /api/v1/voice/{file} 同样需要带头播放。
final class VoiceApi {
    static let shared = VoiceApi()
    static let apiVersion = "v1"
    private let base = "http://127.0.0.1:8787" // 模拟器访问宿主机
    private let session = URLSession.shared
    private var token: String { APIClient.shared.accessToken }

    private init() {}

    func calls(page: Int = 1, completion: @escaping ([CallRecordItem]) -> Void) {
        request("GET", "/api/v1/voice/calls?page=\(page)") { data, _ in
            completion(self.listData(data, as: [CallRecordItem].self))
        }
    }

    func createRoom(name: String, completion: @escaping (Int64?) -> Void) {
        request("POST", "/api/v1/voice/rooms", form: ["name": name]) { data, _ in
            completion(self.dataDict(data)?["room_id"] as? Int64)
        }
    }

    func rooms(page: Int = 1, completion: @escaping ([VoiceRoomItem]) -> Void) {
        request("GET", "/api/v1/voice/rooms?page=\(page)") { data, _ in
            completion(self.listData(data, as: [VoiceRoomItem].self))
        }
    }

    func uploadVoice(fileURL: URL, completion: @escaping ([String: Any]?) -> Void) {
        let boundary = "----SocialBoundary" + String(Int(Date().timeIntervalSince1970 * 1000))
        var body = Data()
        body.append(Data("--\(boundary)\r\n".utf8))
        body.append(Data("Content-Disposition: form-data; name=\"voice\"; filename=\"voice.m4a\"\r\n".utf8))
        body.append(Data("Content-Type: audio/mp4\r\n\r\n".utf8))
        body.append((try? Data(contentsOf: fileURL)) ?? Data())
        body.append(Data("\r\n--\(boundary)--\r\n".utf8))
        request("POST", "/api/v1/im/voice", body: body,
                contentType: "multipart/form-data; boundary=\(boundary)") { data, _ in
            completion(self.dataDict(data))
        }
    }

    private func request(_ method: String, _ path: String, form: [String: String]? = nil,
                         body: Data? = nil, contentType: String = "application/json",
                         completion: @escaping (Data?, Error?) -> Void) {
        var req = URLRequest(url: URL(string: base + path)!)
        req.httpMethod = method
        req.setValue(Self.apiVersion, forHTTPHeaderField: "X-Api-Version")
        if !token.isEmpty {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let form = form {
            let pairs = form.map { "\($0.key)=\($0.value)" }.joined(separator: "&")
            req.httpBody = Data(pairs.utf8)
            req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        } else if let body = body {
            req.httpBody = body
            req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        }
        session.dataTask(with: req) { data, _, error in
            completion(data, error)
        }.resume()
    }

    private func dataDict(_ data: Data?) -> [String: Any]? {
        guard let data = data,
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return root["data"] as? [String: Any]
    }

    private func listData<T: Decodable>(_ data: Data?, as type: T.Type) -> T {
        guard let d = dataDict(data), let list = d["list"],
              let json = try? JSONSerialization.data(withJSONObject: list) else { return [] as! T }
        return (try? JSONDecoder().decode(type, from: json)) ?? ([] as! T)
    }
}

/// 来电 UI 骨架：call_* 信令由 ImClient.onEvent 驱动（媒体面 WebRTC 未实现）。
struct CallView: View {
    @State private var status = "等待来电…"
    @State private var peerId: Int64 = 0
    @State private var canAnswer = false

    var body: some View {
        VStack(spacing: 16) {
            Text(status)
            HStack {
                Button("接听") { send("call_accept") }.disabled(!canAnswer)
                Button("拒绝") { send("call_reject") }.disabled(!canAnswer)
            }
        }
        .onAppear {
            ImClient.shared.onEvent = { type, data in
                DispatchQueue.main.async {
                    switch type {
                    case "call_invite":
                        peerId = data["caller_id"] as? Int64 ?? 0
                        status = "来电 用户#\(peerId)"
                        canAnswer = true
                    case "call_accept": status = "通话中（WebRTC TODO）"
                    case "call_cancel", "call_reject", "call_hangup", "call_failed", "call_timeout":
                        status = "通话结束"
                        canAnswer = false
                    case "call_offer", "call_answer", "call_ice":
                        // TODO 真机联调：WebRTC 媒体面，本里程碑不实现
                        status = "已收到 \(type)（WebRTC TODO）"
                    default: break
                    }
                }
            }
        }
    }

    private func send(_ type: String) {
        ImClient.shared.send(type, data: ["peer_id": peerId])
        canAnswer = false
    }
}

/// 语聊房 UI 骨架：room_* 信令处理（媒体面 WebRTC 未实现）。
struct VoiceRoomView: View {
    @State private var rooms: [VoiceRoomItem] = []
    @State private var status = "房间列表"
    @State private var roomId: Int64 = 0
    @State private var micOn = false

    var body: some View {
        VStack(spacing: 12) {
            Text(status)
            List(rooms) { room in
                Button("\(room.name)  #\(room.id)  在线\(room.onlineCount) 麦\(room.micCount)") {
                    roomId = room.id
                    ImClient.shared.send("room_join", data: ["room_id": room.id])
                    status = "已加入房间 #\(room.id)"
                }
            }
            HStack {
                Button(micOn ? "下麦" : "上麦") {
                    micOn.toggle()
                    ImClient.shared.send(micOn ? "room_up_mic" : "room_down_mic", data: ["room_id": roomId])
                }.disabled(roomId == 0)
                Button("刷新") { load() }
            }
        }
        .onAppear { load() }
    }

    private func load() {
        VoiceApi.shared.rooms { items in
            DispatchQueue.main.async { rooms = items }
        }
    }
}

File: apps/android/app/src/main/java/com/social/app/voice/VoiceApi.kt
// Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
package com.social.app.voice

import com.social.app.api.ApiClient
import java.io.ByteArrayOutputStream
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject

/**
 * 语音功能 HTTP API。所有请求携带 Authorization + X-Api-Version 头。
 * 语音文件播放 GET /api/v1/voice/{file} 同样需要这两个头，客户端必须带头播放。
 */
class VoiceApi(
    private val base: String = BASE,
    private val token: String = ApiClient.accessToken,
) {
    companion object {
        const val API_VERSION = "v1"
        const val BASE = "http://10.0.2.2:8788" // 模拟器访问宿主机
    }

    private val json = Json { ignoreUnknownKeys = true }

    /** 通话历史：GET /api/v1/voice/calls?page= */
    fun calls(page: Int = 1): JsonObject = request("GET", "/api/v1/voice/calls?page=$page")

    /** 创建语聊房：POST /api/v1/voice/rooms {name} */
    fun createRoom(name: String): JsonObject = request("POST", "/api/v1/voice/rooms", mapOf("name" to name))

    /** 开放房间列表：GET /api/v1/voice/rooms?page= */
    fun rooms(page: Int = 1): JsonObject = request("GET", "/api/v1/voice/rooms?page=$page")

    /** 上传语音：multipart field=voice → data.voice_url / data.voice_duration */
    fun uploadVoice(file: File): JsonObject {
        val boundary = "----SocialBoundary" + System.currentTimeMillis()
        val body = ByteArrayOutputStream().apply {
            write("--$boundary\r\n".toByteArray())
            write("Content-Disposition: form-data; name=\"voice\"; filename=\"voice.m4a\"\r\n".toByteArray())
            write("Content-Type: audio/mp4\r\n\r\n".toByteArray())
            write(file.readBytes())
            write("\r\n--$boundary--\r\n".toByteArray())
        }.toByteArray()
        val conn = open("POST", "/api/v1/im/voice")
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
        conn.outputStream.use { it.write(body) }
        return parse(read(conn))
    }

    /** 拼出可带鉴权头播放的完整 URL（base + /api/v1/voice/{file} 形如 voice_url 的值） */
    fun resolve(url: String): String = base + url

    private fun request(method: String, path: String, form: Map<String, String>? = null): JsonObject {
        val conn = open(method, path)
        if (form != null) {
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            conn.outputStream.use { it.write(form.encodeUrlForm().toByteArray()) }
        }
        return parse(read(conn))
    }

    private fun open(method: String, path: String): HttpURLConnection {
        val conn = URL(base + path).openConnection() as HttpURLConnection
        conn.requestMethod = method
        conn.connectTimeout = 10_000
        conn.readTimeout = 15_000
        conn.setRequestProperty("X-Api-Version", API_VERSION)
        if (token.isNotEmpty()) conn.setRequestProperty("Authorization", "Bearer $token")
        return conn
    }

    private fun Map<String, String>.encodeUrlForm(): String =
        entries.joinToString("&") {
            URLEncoder.encode(it.key, "UTF-8") + "=" + URLEncoder.encode(it.value, "UTF-8")
        }

    /** 按状态码读 inputStream 或 errorStream，拿到错误体而非抛异常 */
    private fun read(conn: HttpURLConnection): String =
        try {
            val stream = if (conn.responseCode in 200..299) conn.inputStream else conn.errorStream
            stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        } finally {
            conn.disconnect()
        }

    private fun parse(body: String): JsonObject = json.parseToJsonElement(body).jsonObject
}
